server/src/main/java/org/elasticsearch/index/IndexSettings.java

```diff
@@ -317,12 +321,58 @@
     public static final Setting<Double> FILE_BASED_RECOVERY_THRESHOLD_SETTING
         = Setting.doubleSetting("index.recovery.file_based_threshold", 0.1d, 0.0d, Setting.Property.IndexScope);
 
+    /**
+     * Is the {@code index.mode} enabled? It should only be enbaled if you
+     * pass a jvm parameter or are running a snapshot build.
+     */
+    private static final Boolean TIME_SERIES_MODE_FEATURE_FLAG_REGISTERED;
+
+    static {
+        final String property = System.getProperty("es.index_mode_feature_flag_registered");
+        if (Build.CURRENT.isSnapshot() && property != null) {
+            throw new IllegalArgumentException("es.index_mode_feature_flag_registered is only supported in non-snapshot builds");
+        }
+        TIME_SERIES_MODE_FEATURE_FLAG_REGISTERED = Booleans.parseBoolean(property, null);
+    }
+
+    public static boolean isTimeSeriesModeEnabled() {
+        return Build.CURRENT.isSnapshot() || (TIME_SERIES_MODE_FEATURE_FLAG_REGISTERED != null && TIME_SERIES_MODE_FEATURE_FLAG_REGISTERED);
+    }
+
+    /**
+     * The {@link IndexMode "mode"} of the index.
+     */
+    public static final Setting<IndexMode> MODE = Setting.enumSetting(IndexMode.class,
+        "index.mode",
+        IndexMode.STANDARD,
+        new Setting.Validator<IndexMode>() {
+            @Override
+            public void validate(IndexMode value) {}
+
+            @Override
+            public void validate(IndexMode value, Map<Setting<?>, Object> settings) {
+                value.validateWithOtherSettings(settings);
+            }
+
+            @Override
+            public Iterator<Setting<?>> settings() {
+                return IndexMode.VALIDATE_WITH_SETTINGS.iterator();
+            }
+        },
+        Property.IndexScope
```

Comment:
This should probably also have `Property.Final`.

Reply:
Indeed `Property.Final` seems like that way to go here.

Reply:
Should `INDEX_ROUTING_PARTITION_SIZE_SETTING` be Final as well? It is not marked Final but it seems like it shouldn't be set after index creation.

Reply:
I opened #77669 to talk about that.